moxin-backend/src/backend_impls.rs

```diff
@@ -0,0 +1,716 @@
+        chat_completion_message: Option<Vec<u8>>,
+        pub token_tx: Option<Sender<anyhow::Result<ChatResponse>>>,
+        pub load_model_state: Option<(
+            FileID,
+            LoadModelOptions,
+            Sender<anyhow::Result<LoadModelResponse>>,
+        )>,
+    }
+
+    impl ChatBotUi {
+        pub fn new(
+            request_rx: Receiver<(ChatRequestData, Sender<anyhow::Result<ChatResponse>>)>,
+            load_module_req: (
+                FileID,
+                LoadModelOptions,
+                Sender<anyhow::Result<LoadModelResponse>>,
+            ),
+        ) -> Self {
+            Self {
+                request_rx,
+                token_tx: None,
+                current_req: std::io::Cursor::new(vec![]),
+                load_model_state: Some(load_module_req),
+                chat_completion_message: None,
+            }
+        }
+
+        fn init_request(&mut self) -> Result<(), ()> {
+            if let Ok((req, tx)) = self.request_rx.recv() {
+                // Init current_req
+                if !req.stream.unwrap_or_default() {
+                    self.chat_completion_message = Some(Vec::with_capacity(
+                        (req.max_tokens.unwrap_or(512) * 8) as usize,
+                    ))
+                }
+                *self.current_req.get_mut() = serde_json::to_vec(&req).unwrap();
+                self.current_req.set_position(0);
+                self.token_tx = Some(tx);
+                Ok(())
+            } else {
+                Err(())
+            }
+        }
+
+        pub fn read_data(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+            let n = self.current_req.read(buf)?;
+            if n == 0 {
+                self.current_req.get_mut().clear();
+                self.current_req.set_position(0);
+            }
+            Ok(n)
+        }
+
+        fn send_output(&mut self, output: Result<&[u8], TokenError>) -> bool {
+            if self.token_tx.is_none() {
+                return false;
+            }
+
+            match (output, &mut self.chat_completion_message) {
+                (Ok(token), Some(chat_completion_message)) => {
+                    chat_completion_message.extend_from_slice(token);
+                }
+                (Ok(token), None) => {
+                    let _ =
+                        self.token_tx
+                            .as_mut()
+                            .unwrap()
+                            .send(Ok(ChatResponse::ChatResponseChunk(ChatResponseChunkData {
+                                id: String::new(),
+                                choices: vec![ChunkChoiceData {
+                                    finish_reason: None,
+                                    index: 0,
+                                    delta: MessageData {
+                                        content: String::from_utf8_lossy(token).to_string(),
```

Comment:
I noticed the first chunks comes with some blank spaces at the beginning. For instance, these are the two first chunks I'm receiving in one run:

```
        ChunkChoiceData {
            finish_reason: None,
            index: 0,
            delta: MessageData {
                content: " ",
                role: Assistant,
            },
            logprobs: None,
        },
        
        ChunkChoiceData {
            finish_reason: None,
            index: 0,
            delta: MessageData {
                content: " Gra",
                role: Assistant,
            },
            logprobs: None,
        },
```

Do you know the reason?

In the meanwhile, I will trim the answer in the UI to make sure we don't show extra whitespaces.

Reply:
Actually, I'm not sure what the reason is either; in some cases, the model's output does start with a space token. I think perhaps a solution would be to ignore these space tokens when displaying them until the first non-space token appears

Reply:
Sure, no problem at all. Thanks for confirming it is an issue we should address in the frontend.